### PR TITLE
feat(terraform): Implement terraform providers

### LINKED ISSUE #4267: Support Terraform Providers
**What would you like Renovate to be able to do?**
It would be great if the Terraform support could be widened to include `providers`  in addition to `modules`.

**Additional context**
- [Renovate feature request (successfully implemented) for Terraform, which added support for Terraform _modules_](https://github.com/renovatebot/renovate/issues/1860)
- [Terraform docs on providers](https://www.terraform.io/docs/providers/).
- [Provider versioning syntax (looks like standard `hashicorp` style?](https://www.terraform.io/docs/configuration/providers.html#version-provider-versions)
- [Sample available versions for the "github" provider](https://releases.hashicorp.com/terraform-provider-github/)

### PR DESCRIPTION
**Pull request by secustor**
Implements a manager and datasource for Terraform providers.    

The code is based upon the `terraform` manager and reuses the versioning system of `hashicorp`

I have tested it against a local Gitlab instance with the providers for Azure, AWS and Kubernetes.   

Please be aware that this is my first contribution in Typescript so I suggest a more thoroughly review than normal. 

Closes #4267

### CODE REVIEW

**Comment by rarkins** [comment]
@secustor thanks for your PR. Did you consider or disregard whether this functionality could be incorporated into the existing terraform managers and/or datasources? I'm not saying they definitely should be, but I'm wondering if you had found any reason not to.

**Comment by secustor** [comment]
Thanks for the fast answer!

I have looked into it a bit, but the problem I see with an integrated approach is the differentiation between the modules and providers.

Reasons I decided to separate the providers: 
-  `DepType` is already used in `terraform`.  Is `lookupType` the variable to use here? 
-  Modules are served under a different URL and have its own namespace
-  Easier integration of more advanced features like changelogs or naming including the alias names, but these have been out of scope for this MVC. 

We could also rewrite `terraform` to use other variables then the dependency type. IMHO would this make the `terraform` manager even harder to read then it is at the moment. 

I am open for suggestions here!

**Comment by rarkins** [comment]
It's possible to have different datasources but the same manager to extract. That would be the first thing I'd explore. i.e.

- keep your new terraform-provider datasource
- enhance the existing terraform extract and update functions to recognize provider references 

You can use `depTypes` instead of `depType` if you feel that dependencies need to be classified in more than one way.

File: lib/manager/terraform-provider/extract.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      if (!isValid(dep.currentValue)) {
+        dep.skipReason = 'unsupported-version';
+      }
+    } else if (!dep.skipReason) {
+      dep.skipReason = 'no-version';
+    }
+
+    delete dep.versionLine;
+    /* eslint-enable no-param-reassign */
+  });
```
Comment (ViceIce) [inline lib/manager/terraform-provider/extract.ts]: This should be integrated to the existing loop.

**Review by viceice — COMMENTED** [review]
Code looks good, but i'm with @rarkins to unify the managers

**Comment by secustor** [comment]
@rarkins @ViceIce 
I have integrated the logic of the `terraform-provider` manager into the existing one.  

Please review again.

**Comment by rarkins** [comment]
I like the way this is looking. Thanks @secustor for making the changes. Do I understand correctly that the `lib/manager/terraform/update.ts` code didn't even need updating this way?

**Comment by secustor** [comment]
@rarkins Yes this is correct. As both the provider and the module definition are using the `version` key, it is enough to define in the manager the `depType` variable as `terraform`

**Comment by rarkins** [comment]
Great, but let's also have some update tests as well, to make sure we don't break anything in future.

**Comment by secustor** [comment]
@rarkins I have added two tests for `update.ts`

File: lib/datasource/terraform-provider/index.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      versions: {},
+      releases: null,
+    };
+    if (res.source) {
+      dep.sourceUrl = parse(res.source);
+    }
+    dep.releases = res.versions.map(version => ({
+      version,
+    }));
+    if (pkgUrl.startsWith('https://registry.terraform.io/')) {
```
Comment (rarkins) [inline lib/datasource/terraform-provider/index.ts]: Isn't this always true?

File: lib/manager/terraform/extract.ts [hunk 3]
```diff
@@ -2,83 +2,138 @@ import { logger } from '../../logger';
 import { isValid, isVersion } from '../../versioning/hashicorp';
 import { PackageDependency, PackageFile } from '../common';
 
+export enum TerraformDependencyTypes {
+  unknown = 'unknown',
+  module = 'module',
+  provider = 'provider',
+}
+
+export function getTerraformDependencyType(
```
Comment (rarkins) [inline lib/manager/terraform/extract.ts]: Isn't this function a little overkill? We only have two types that our regex matches, so it's impossible to match default too, right?

File: lib/datasource/terraform-provider/index.ts (hunk 2 above)
Comment (secustor) [reply lib/datasource/terraform-provider/index.ts]: Yeah you right. This is a left over from the `terraform` datasource.

File: lib/manager/terraform/extract.ts (hunk 3 above)
Comment (secustor) [reply lib/manager/terraform/extract.ts]: IMHO I prefer to use enums for such kind of tasks as it is easier to read and it is a build in fail safe for errors.     
From a performance point of view there should be no difference.

**Pull request merged**